stdlib/dynarray.ml

```diff
@@ -267,6 +275,41 @@ end = struct
         done;
       end
 
+    let blit src src_dummy src_pos dst dst_dummy dst_pos ~len =
+      if src_dummy == dst_dummy then
+        Array.blit src src_pos dst dst_pos len
+      else begin
+        if len < 0
+           || src_pos < 0
+           || src_pos + len < 0 (* overflow check *)
+           || src_pos + len > Array.length src
+           || dst_pos < 0
+           || dst_pos + len < 0 (* overflow check *)
+           || dst_pos + len > Array.length dst
+        then
+          (* We assume that the caller has already checked this and
+             will raise a proper error. The check here is only for
+             memory safety, it should not be reached and it is okay if
+             the error is uninformative. *)
+          invalid_arg "dummy";
+        (* In case of self-blitting with overlap we must be careful about
+           the copying order -- we might overwrite an element we need later on.
+           Same logic as runtime/array.c *)
+        if src == dst && src_pos <= dst_pos then
+          (* copy in descending order *)
+          if src_pos = dst_pos then ()
+          else for i = len - 1 downto 0 do
+            Array.unsafe_set dst (dst_pos + i)
+              (Array.unsafe_get src (src_pos + i));
+          done
```

Comment:
is there ever a case where we're self-blitting but dummies differ?

Reply:
Ah, good catch! If we already know that the dummies differ (we checked this above), then we know that we cannot be in the self-blit scenario, as the backing arrays of distinct dynarrays are always distinct. I wrote the same-dummy test above strictly as an optimization over a correct-on-its-own implementation, but in fact I could simplify the implementation here based on this precondition.

(I'm not sure what you meant about the stamp types, but in any case they are sort of a distraction here: outside the Dummy signature the function is typed to accept arrays with different stamp types, so we don't know much about the stamps.)

Reply:
> I'm not sure what you meant about the stamp types

I was trying to find a contradiction in my reasoning. What I meant was that inside the positive branch of this if condition, which I believe is unreachable in all cases, those two facts are true:  `dummy1 != dummy2`, `arr1 == arr2`.
but if `arr1 == arr2`, it must be the case for me as a caller that I can't pass `blit` any other dummy for either of them, because the stamp types won't match with their respective array (and by extension the other dummy). e.g.
```ocaml
let (Pack a) = Dynarray.create() in
let (Pack b) = Dynarray.create() in
Dynarray.Dummy.Array.blit a.arr a.dummy 0 a.arr b.dummy 0 ~len:0
```
so it must be the case that the dummies are in fact equal.
this is built on my assumption that `'stamp` ensures the following implication `stamp1 = stamp2 -> (dummy1 == dummy 2) = true`.. and I wasn't really sure of that.

I was actually trying to find a way to ever safely hit a case where dummies aren't equal (in stamp or value) but the arrays are (in value), and I found that this is only the case if the arrays are empty (and that's only because the empty array has a stable out of heap repr like the empty list etc, so it's a harmless -- and useless -- blit).

Reply:
Indeed, the question is to find a simple explanation for why `a.dummy != b.dummy` implies either `a.arr != b.arr` or `a.arr == b.arr == [||]`.  I think that the reasoning is that `a == b` implies `a.dummy == b.dummy` (obviously) and that `a != b` implies `a.arr != b.arr` or `a.arr == b.arr == [||]`. The second part of this argument, that distinct dynarrays have distinct backing arrays, is a global invariant of the whole Dynarray module, which can be checked by looking at all functions to ensure that they respect this invariant: all initializations or writes to the `arr` field of a dynarray use an empty or fresh/separated backing array.

Reply:
IIUC this is as a discussion about a possible simplification/optimization and not an incorrectness, so my approval still stands and you can merge as-is, unless you want to update the PR with further improvements.

Reply:
I ended up implementing the small simplification, which makes the code shorter and easier to read. I also rebased the PR on the current trunk.